File: data structures and algorithms/python/08_038.py
def count_convoys(target, positions, speeds):

    # Pair positions with speeds and sort by position
    # in descending order
    cars = sorted(zip(positions, speeds), reverse=True)
    
    convoys = 0
    prev_time = -1
    
    for position, speed in cars:
        # Calculate time to reach target
        time = (target - position) / speed
        
        # If this vehicle takes longer to reach target,
        # it forms a new convoy
        if time > prev_time:
            convoys += 1
            prev_time = time
            
    return convoys

File: data structures and algorithms/cs/08_038.cs
using System;
using System.Linq;

public class ConvoyCounter {
    public static int CountConvoys(int target, int[] positions, int[] speeds) {
        int n = positions.Length;
        var cars = new double[n][];

        for (int i = 0; i < n; i++) {
            cars[i] = new double[] { positions[i], speeds[i] };
        }

        Array.Sort(cars, (a, b) => b[0].CompareTo(a[0]));

        int convoys = 0;
        double prevTime = -1;

        for (int i = 0; i < n; i++) {
            double position = cars[i][0];
            double speed = cars[i][1];
            double time = (target - position) / speed;

            if (time > prevTime) {
                convoys++;
                prevTime = time;
            }
        }

        return convoys;
    }
}